app/services/course/assessment/answer/programming_auto_grading_service.rb

```diff
@@ -54,32 +55,56 @@ def evaluate_package(question, package)
   #   graded.
   # @param [Course::Assessment::ProgrammingEvaluationService::Result] evaluation_result The
   #   result of evaluating the package.
-  # @param [Boolean] ignore_evaluation If set to true, the evaluation results will be ignored when
-  #   determine the correctness of the answer.
-  # @return [Array<(Boolean, Integer, Course::Assessment::Answer::ProgrammingAutoGrading)>] The
-  #   correct status, grade and the programming auto grading record.
-  def build_result(question, evaluation_result, ignore_evaluation:)
+  # @param [Array<String>] graded_test_case_types The types of test cases counted
+  #   towards grade/exp calculation
+  # @return [Array<(Boolean, Integer, Course::Assessment::Answer::ProgrammingAutoGrading), Integer>]
+  #   The correctness apparent to student ('True' if answer passes public and private test
+  #   cases), grade, the programming auto grading record, and the evaluation result's id.
+  def build_result(question, evaluation_result, graded_test_case_types:)
+    auto_grading = build_auto_grading(question, evaluation_result)
+    graded_test_count = question.test_cases.where(test_case_type: graded_test_case_types).size
+    number_correct = auto_grading.test_results.
+                     select { |r| graded_test_case_types.include?(r.test_case.test_case_type) }.
+                     map(&:passed?).count(true)
```

Comment:
`select { |r| graded_test_case_types.include?(r.test_case.test_case_type) && r.passed? }.count`

Same for the line in `check_correctness`.

Actually you can abstract it out into a `def count_correct(auto_grading, test_case_types)` function.